### PR TITLE
OOP school library: unit tests

### PR DESCRIPTION
**Pull request by Mo7amedSMahdi**
# OOP school library: unit tests

## Description
In this project, I added unit tests to your codebase, specifically to main classes.

### Project requirements
- Add unit tests to your classes:
  - `Book`
  - `Person`
  - `Student`
  - `Teacher`
  - `Classroom`
  - `Rental`
  - all classes that you created while implementing the Decorator design pattern

### CODE REVIEW

File: spec/person_spec.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    end
+
+    it 'Person id should be 1' do
+      expect(person.id).to eq(1)
+    end
+
+    it 'Person parent_permission should be true' do
+      expect(person.parent_permission).to eq(true)
+    end
+  end
```
Comment (t-yanick) [inline spec/person_spec.rb:38]: - [x] Good job with test coverage so, but since we are aiming for a 100% test coverage, it will be nice to cover all our methods in all our classes with tests, for example in this `Person class`, the following tests could be added; 
- Test for the `add_rental method`: 
```
  context "When adding a new rentals" do

    it "The add_rental method returns a rental and adds its to the book" do
      book = Book.new("Title", "Author")
      person = Person.new(24, "Bob")

      rental = person.add_rental(book, "2021-05-04")

      expect(person.rentals).to eql [rental]
    end
  end
```
- The for `checking_permission method`: 
```ruby
  context "When checking permissions" do
    
    it "The can_use_services? method returns false if under age and no parent permission" do
      person = Person.new(14, "Rose", false)
      expect(person.can_use_services?).to be false
    end
    
    it "The can_use_services? method returns true if over age" do
      person = Person.new(19, "Sam", false)
      expect(person.can_use_services?).to be true
    end
    
    it "The can_use_services? method returns true if has parent permission" do
      person = Person.new(14, "Rose", true)
      expect(person.can_use_services?).to be true
    end
  end
```
- Test for `validating_name method`: 
```ruby
  context "When validating name" do

    it "The validate_name method corrects the name" do
      person = Person.new(19, "alongnameiscorrected")
      person.validate_name()
      expect(person.name).to eql "Alongnamei"
    end
  end
```
#### Please it's cover all our tests in all our classes with tests and demonstrated about, it will also be nice to check for positive and negative cases. Checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/rspec_unit_test.md) for more clarification.

**Review by t-yanick — CHANGES_REQUESTED** [review]
Hi @Mo7amedSMahdi, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/person_spec.rb (hunk 1 above)
Comment (Mo7amedSMahdi) [reply spec/person_spec.rb:38]: @t-yanick Thank you for your feedback, Sorry if I miss that because I didn't know that we aiming for 100% coverage in the test, I will work on this fix as soon as possible.

**Review by aimemalaika — APPROVED** [review]
Hi @Mo7amedSMahdi, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

133591850-fdcb7dd3-dd7a-4274-81e7-0e59434148f7

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
